File: inspector.py
"""
Core inspection logic.
Sends frames + audio to Gemma 4 12B via Ollama in a single multimodal prompt.

Key insight: because Gemma 4 12B has no separate encoders, audio and video tokens
share the same RoPE positional space. The model can reason about temporal alignment
between what it sees and what it hears without any intermediate representation.
"""

import json
import ollama
from extract import extract_frames, extract_audio_segment, get_video_duration, audio_to_base64


MODEL = "gemma4:12b"

SYSTEM_PROMPT = """You are a temporal sync analyzer. You receive video frames (with timestamps)
and the corresponding audio from the same time window.

Your job: determine whether the audio and video are temporally synchronized.

Focus on:
1. Lip sync — do mouth movements match the phonemes/words in the audio?
2. Action sync — do physical actions (clapping, typing, door closing) match audio events?
3. Emotional sync — does facial expression match vocal tone?

Respond ONLY with valid JSON in this exact schema:
{
  "in_sync": boolean,
  "confidence": "high" | "medium" | "low",
  "desync_detected_at_s": number | null,
  "desync_type": "lip_sync" | "action_sync" | "emotional_sync" | null,
  "estimated_offset_ms": number | null,
  "reasoning": "one concise sentence",
  "suspicious_frames": [timestamp_s, ...]
}"""


def build_prompt(frames: list[dict], window_start_s: float, window_end_s: float) -> str:
    frame_list = "\n".join(
        f"  - Frame at t={f['timestamp_s']:.1f}s" for f in frames
    )
    return (
        f"Analyze this {window_end_s - window_start_s:.0f}s video segment "
        f"(t={window_start_s:.1f}s to t={window_end_s:.1f}s).\n\n"
        f"Frames provided:\n{frame_list}\n\n"
        "The audio for this exact window is also provided. "
        "Determine if audio and video are in sync."
    )


def inspect_window(
    video_path: str,
    window_start_s: float = 0.0,
    window_duration_s: float = 30.0,
    fps: int = 1,
) -> dict:
    """
    Inspect a single time window of a video for A/V sync.
    Returns the parsed JSON result from Gemma 4 12B.
    """
    window_end_s = window_start_s + window_duration_s

    # Extract frames for this window
    frames = extract_frames(video_path, fps=fps, max_frames=int(window_duration_s * fps))
    # Filter to window
    frames = [f for f in frames if window_start_s <= f["timestamp_s"] < window_end_s]

    # Extract audio for this window (16kHz WAV — Gemma's native format)
    audio_path = extract_audio_segment(video_path, window_start_s, window_duration_s)
    audio_b64 = audio_to_base64(audio_path)

    # Build multimodal message: images + audio + text in one prompt
    # Gemma 4 12B processes all modalities through the same backbone
    images = [f["b64"] for f in frames]

    prompt_text = build_prompt(frames, window_start_s, window_end_s)

    response = ollama.chat(
        model=MODEL,
        messages=[
            {"role": "system", "content": SYSTEM_PROMPT},
            {
                "role": "user",
                "content": prompt_text,
                "images": images,
            },
        ],
        options={"temperature": 0.1},  # low temp for deterministic analysis
    )

    raw = response["message"]["content"].strip()

    # Strip markdown code fences if present
    if raw.startswith("```"):
        raw = raw.split("```")[1]
        if raw.startswith("json"):
            raw = raw[4:]

    try:
        result = json.loads(raw)
    except json.JSONDecodeError:
        result = {
            "in_sync": None,
            "confidence": "low",
            "desync_detected_at_s": None,
            "desync_type": None,
            "estimated_offset_ms": None,
            "reasoning": raw[:300],
            "suspicious_frames": [],
            "parse_error": True,
        }

    result["window_start_s"] = window_start_s
    result["window_end_s"] = window_end_s
    return result


def inspect_full_video(
    video_path: str,
    window_duration_s: float = 30.0,
    fps: int = 1,
    progress_cb=None,
) -> list[dict]:
    """
    Slide a window across the full video and inspect each segment.
    Returns list of per-window results.
    """
    duration = get_video_duration(video_path)
    results = []
    start = 0.0
    window_idx = 0
    total_windows = int(duration / window_duration_s) + 1

    while start < duration:
        actual_duration = min(window_duration_s, duration - start)
        if actual_duration < 5:
            break

        if progress_cb:
            progress_cb(window_idx / total_windows, f"Analyzing t={start:.0f}s–{start+actual_duration:.0f}s…")

        result = inspect_window(video_path, start, actual_duration, fps)
        results.append(result)
        start += window_duration_s
        window_idx += 1

    return results


def summarize_results(results: list[dict]) -> dict:
    """Aggregate per-window results into a video-level summary."""
    desync_windows = [r for r in results if not r.get("in_sync")]
    total = len(results)

    return {
        "total_windows": total,
        "desync_windows": len(desync_windows),
        "sync_score": round((total - len(desync_windows)) / max(total, 1) * 100, 1),
        "verdict": "IN SYNC" if len(desync_windows) == 0 else (
            "MINOR DESYNC" if len(desync_windows) / max(total, 1) < 0.3 else "SIGNIFICANT DESYNC"
        ),
        "desync_timestamps": [
            r["desync_detected_at_s"] for r in desync_windows
            if r.get("desync_detected_at_s") is not None
        ],
        "windows": results,
    }

File: internal/inspector/inspector.go
// Package inspector drives the audio-visual sync analysis.
// It slides a window across the video, sends frames + audio to Gemma 4
// in a single multimodal prompt, and parses the structured JSON result.
package inspector

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/Siddhant-K-code/lipsync-check/internal/extract"
	"github.com/Siddhant-K-code/lipsync-check/internal/ollama"
)

const systemPrompt = `You are an audio-visual sync analyzer. You receive video frames (with timestamps) and the corresponding audio from the same time window.

Your job: determine whether the audio and video are synchronized.

Focus on:
1. Lip sync — do mouth movements match the phonemes/words in the audio?
2. Action sync — do physical actions (clapping, typing, door closing) match audio events?
3. Emotional sync — does facial expression match vocal tone?

Respond ONLY with valid JSON in this exact schema:
{
  "in_sync": boolean,
  "confidence": "high" | "medium" | "low",
  "desync_detected_at_s": number | null,
  "desync_type": "lip_sync" | "action_sync" | "emotional_sync" | null,
  "estimated_offset_ms": number | null,
  "reasoning": "one concise sentence",
  "suspicious_frames": [timestamp_s, ...]
}`

// WindowResult holds the analysis result for a single time window.
type WindowResult struct {
	WindowStartS float64  `json:"window_start_s"`
	WindowEndS   float64  `json:"window_end_s"`
	InSync       *bool    `json:"in_sync"`
	Confidence   string   `json:"confidence"`
	DesyncAtS    *float64 `json:"desync_detected_at_s"`
	DesyncType   *string  `json:"desync_type"`
	OffsetMS     *float64 `json:"estimated_offset_ms"`
	Reasoning    string   `json:"reasoning"`
	SuspFrames   []float64 `json:"suspicious_frames"`
	ParseError   bool     `json:"parse_error,omitempty"`
	RawResponse  string   `json:"raw_response,omitempty"`
}

// Summary aggregates all window results into a video-level verdict.
type Summary struct {
	TotalWindows    int             `json:"total_windows"`
	DesyncWindows   int             `json:"desync_windows"`
	SyncScore       float64         `json:"sync_score"`
	Verdict         string          `json:"verdict"`
	DesyncTimestamps []float64      `json:"desync_timestamps"`
	Windows         []WindowResult  `json:"windows"`
}

// Config holds inspector parameters.
type Config struct {
	OllamaHost string
	Model      string
	FPS        int
	WindowS    float64
	MaxFrames  int
}

// DefaultConfig returns sensible defaults.
func DefaultConfig() Config {
	return Config{
		OllamaHost: "http://localhost:11434",
		Model:      "gemma4:e4b",
		FPS:        1,
		WindowS:    30.0,
		MaxFrames:  60,
	}
}

// ProgressFunc is called after each window with (windowIndex, totalWindows, startS).
type ProgressFunc func(idx, total int, startS float64)

// InspectVideo slides a window across the full video and inspects each segment.
func InspectVideo(ctx context.Context, videoPath string, cfg Config, progress ProgressFunc) (*Summary, error) {
	client := ollama.New(cfg.OllamaHost, cfg.Model)

	if err := client.Ping(ctx); err != nil {
		return nil, fmt.Errorf("ollama not reachable: %w", err)
	}

	duration, err := extract.Duration(videoPath)
	if err != nil {
		return nil, fmt.Errorf("get duration: %w", err)
	}

	var results []WindowResult
	total := int(duration/cfg.WindowS) + 1
	idx := 0

	for start := 0.0; start < duration; start += cfg.WindowS {
		windowDur := cfg.WindowS
		if start+windowDur > duration {
			windowDur = duration - start
		}
		if windowDur < 5 {
			break
		}

		if progress != nil {
			progress(idx, total, start)
		}

		result, err := inspectWindow(ctx, client, videoPath, start, windowDur, cfg)
		if err != nil {
			// Non-fatal: record the error and continue
			t := true
			result = WindowResult{
				WindowStartS: start,
				WindowEndS:   start + windowDur,
				InSync:       &t,
				Confidence:   "low",
				Reasoning:    fmt.Sprintf("analysis failed: %v", err),
				ParseError:   true,
			}
		}
		results = append(results, result)
		idx++
	}

	return summarize(results), nil
}

// InspectWindow analyzes a single time window.
func InspectWindow(ctx context.Context, videoPath string, startS, durationS float64, cfg Config) (*WindowResult, error) {
	client := ollama.New(cfg.OllamaHost, cfg.Model)
	r, err := inspectWindow(ctx, client, videoPath, startS, durationS, cfg)
	if err != nil {
		return nil, err
	}
	return &r, nil
}

func inspectWindow(ctx context.Context, client *ollama.Client, videoPath string, startS, durationS float64, cfg Config) (WindowResult, error) {
	result := WindowResult{
		WindowStartS: startS,
		WindowEndS:   startS + durationS,
	}

	// Extract frames
	frames, err := extract.ExtractFrames(videoPath, startS, durationS, cfg.FPS, cfg.MaxFrames)
	if err != nil {
		return result, fmt.Errorf("extract frames: %w", err)
	}
	defer func() {
		for _, f := range frames {
			extract.Cleanup(f.Path)
		}
	}()

	// Extract audio (16kHz WAV — Gemma 4 native)
	wavPath, err := extract.ExtractAudio(videoPath, startS, durationS)
	if err != nil {
		return result, fmt.Errorf("extract audio: %w", err)
	}
	defer extract.Cleanup(wavPath)

	audioB64, err := extract.AudioToBase64(wavPath)
	if err != nil {
		return result, fmt.Errorf("encode audio: %w", err)
	}

	// Build image list: frames + audio in one slice
	// Gemma 4's encoder-free design handles both through the same token projection
	images := make([]string, 0, len(frames)+1)
	for _, f := range frames {
		images = append(images, f.Base64JPEG)
	}
	images = append(images, audioB64)

	// Build prompt
	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("Analyze this %.0fs video segment (t=%.1fs to t=%.1fs).\n\nFrames provided:\n",
		durationS, startS, startS+durationS))
	for _, f := range frames {
		sb.WriteString(fmt.Sprintf("  - Frame at t=%.1fs\n", f.TimestampS))
	}
	sb.WriteString("\nThe audio for this exact window is also provided. Determine if audio and video are in sync.")

	raw, err := client.Chat(ctx, systemPrompt, sb.String(), images)
	if err != nil {
		return result, fmt.Errorf("ollama chat: %w", err)
	}

	// Strip markdown fences if present
	cleaned := strings.TrimSpace(raw)
	if strings.HasPrefix(cleaned, "```") {
		parts := strings.SplitN(cleaned, "\n", 2)
		if len(parts) == 2 {
			cleaned = parts[1]
		}
		cleaned = strings.TrimSuffix(strings.TrimSpace(cleaned), "```")
	}

	if err := json.Unmarshal([]byte(cleaned), &result); err != nil {
		result.ParseError = true
		result.RawResponse = raw[:min(len(raw), 400)]
		result.Reasoning = fmt.Sprintf("parse error: %v", err)
	}

	result.WindowStartS = startS
	result.WindowEndS = startS + durationS
	return result, nil
}

func summarize(results []WindowResult) *Summary {
	var desyncWindows int
	var desyncTS []float64

	for _, r := range results {
		if r.InSync != nil && !*r.InSync {
			desyncWindows++
			if r.DesyncAtS != nil {
				desyncTS = append(desyncTS, *r.DesyncAtS)
			}
		}
	}

	total := len(results)
	score := 0.0
	if total > 0 {
		score = float64(total-desyncWindows) / float64(total) * 100
	}

	verdict := "IN SYNC"
	if desyncWindows > 0 {
		ratio := float64(desyncWindows) / float64(max(total, 1))
		if ratio >= 0.3 {
			verdict = "SIGNIFICANT DESYNC"
		} else {
			verdict = "MINOR DESYNC"
		}
	}

	return &Summary{
		TotalWindows:     total,
		DesyncWindows:    desyncWindows,
		SyncScore:        score,
		Verdict:          verdict,
		DesyncTimestamps: desyncTS,
		Windows:          results,
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
